### PR TITLE
ci: bump Mac OS 12 and XCode 14 in CI

### PR DESCRIPTION
**Pull request by rerorero**
SSIA, see https://github.com/appium/appium-xcuitest-driver/pull/1478 for the background.

ref macos-12 https://github.com/actions/runner-images/blob/main/images/macos/macos-12-Readme.md

### CODE REVIEW

File: ci-jobs/pr-validation-build.yml [hunk 1]
```diff
@@ -13,20 +13,20 @@ stages:
...
       iosVersion: 15.2
-      xcodeVersion: 13.2
+      xcodeVersion: 14.0
       deviceName: "iPhone Simulator"
       #tvosVersion: 13.4
       #tvosName: AppleTV_13_4
       #tvosDeviceName: "Apple TV"
       skipTvOs: True
-      vmImage: 'macOS-11'
+      vmImage: 'macOS-12'
```
Comment (rerorero) [inline ci-jobs/pr-validation-build.yml:22]: https://github.com/actions/runner-images/issues/6384
macos-latest still points to 11

File: ci-jobs/pr-validation-build.yml [hunk 2]
```diff
@@ -11,22 +11,19 @@ stages:
   jobs:
   - template: ./templates/xcuitest-e2e-template.yml
     parameters:
-      name: iPhoneX_15_2
-      iosVersion: 15.2
-      xcodeVersion: 13.2
-      deviceName: "iPhone Simulator"
```
Comment (rerorero) [inline ci-jobs/pr-validation-build.yml:17]: How do you decide 'generic' simulator? https://github.com/appium/appium-xcuitest-driver/blob/master/lib/ios-generic-simulators.js
Apparently, the CI env no longer has `iPhone X` for iOS 16.2 https://dev.azure.com/AppiumCI/Appium%20CI/_build/results?buildId=24430&view=logs&j=19e62f86-4150-54d4-d8db-ae5426c0dd0e&t=edcc93af-4c73-59ad-5b41-a879e5dae5b2&l=116

Comment (mykola-mokhnach) [reply ci-jobs/pr-validation-build.yml:17]: there are two ways to do it:
either you hardcode the device model and it gets outdated after a while when Apple stops supporting it or fetch the model dynamically by parsing the output of `xcrun simctl list` in the CI script

Comment (rerorero) [reply ci-jobs/pr-validation-build.yml:17]: I was just wondering if there is a policy to pick which device to use for tests. If you don't mind, I'm going to hard code iPhone 12 & 13

**Comment by rerorero** [comment]
hmm.. I'll check the CI result 👀  Perhaps  I can revisit this after my new year holidays. Please bear with me.

File: ci-jobs/pr-validation-build.yml (hunk 2 above)
Comment (KazuCocoa) [reply ci-jobs/pr-validation-build.yml:17]: We don't have it right now. Some test cases potentially depend on the device resolution etc, but as XCUITest driver' we don't have such a policy. (in WDA, we'd like to run tests in various combinations of Xcode/OS versions)

Comment (rerorero) [reply ci-jobs/pr-validation-build.yml:17]: I see, thank you for clarifying!

**Comment by mykola-mokhnach** [comment]
seems like there are still some flacky tests. Perhaps, we need to further increase wda build timeout (currently 480 seconds) because of CI slowness?

**Comment by rerorero** [comment]
[Now the viewport screenshot test is failing](https://dev.azure.com/AppiumCI/Appium%20CI/_build/results?buildId=24477&view=logs&j=7ab5344b-cfd2-53e8-5e76-b5822bec072b&t=62dc0973-5593-51ea-5478-b1d14625d345&l=33497). I confirmed I can reproduce it on local as well.
Here are the images obtained using takeScreenshot and viewportScreenshot. The viewport image is cropped like the surrounding area is trimmed. 
full
viewport

This is because[ the size returned by getWindowSize](https://github.com/appium/appium-xcuitest-driver/blob/166d4e3425c92b041f835df988a071e55a0f2862/lib/commands/screenshots.js#L71) is different from the size of the screenshot.
[getWindowSize() returns `width:375, height:812` ](https://dev.azure.com/AppiumCI/Appium%20CI/_build/results?buildId=24477&view=logs&j=19d171a9-dbae-5036-c81d-2485d7f00528&t=f509ce69-fe42-5b92-2604-61038d9323ed&l=6881)but takeScreenshot returns an image with the size `width: 1179, height: 2556` in pixel i.e. `width:393, height:852`
I'm not sure if this is expected behavior 🤔

**Comment by mykola-mokhnach** [comment]
yes, this is expected. The screenshot size is in pixels and getWindowSize returns logical units. Depending on the device model the scale factor between units and pixels could be 2x or 3x. The actual scale factor value could be retrieved from https://github.com/appium/appium-xcuitest-driver#mobile-devicescreeninfo

**Comment by rerorero** [comment]
I mean I wonder if it's expected that getWindowSize() can return the different size of the image returned by the takeScreenshot() (with multiplied by factor)
With `iPhone X` + iOS 14.4 , getWindowSize() returns `w=375 h=812` and takeScreenshot() returns an image with `w=1125, h=2436`. This passes the test because 812 * 3 = 2436.
But with `iPhone 12` + iOS 14.4, getWindowSize() returns `w=375 h=812` and takeScreenshot() returns `w=1170 h=2532`. This causes the test failure as 812 * 3 ≠ 2532.

**Comment by mykola-mokhnach** [comment]
It could be that the phone has some additional area where system-specific UI components are living. And this area does not get added into the returned screen metrics because user apps anyway cannot use it

File: test/functional/basic/basic-e2e-specs.js [hunk 3]
```diff
@@ -203,9 +203,10 @@ describe('XCUITestDriver - basics -', function () {
       await B.promisify(fullImg.parse, {context: fullImg})(fullB64);
       const viewImg = new PNG({filterType: 4});
       await B.promisify(viewImg.parse, {context: viewImg})(viewB64);
-      fullImg.height.should.eql(viewImg.height + Math.round(pixelRatio * statBarHeight));
+      // Viewport size can be smaller than the full image size + status bar on some devices.
+      fullImg.height.should.be.gte(viewImg.height + Math.round(pixelRatio * statBarHeight));
```
Comment (rerorero) [inline test/functional/basic/basic-e2e-specs.js:207]: Okay, I changed the condition https://github.com/appium/appium-xcuitest-driver/pull/1480#issuecomment-1371063721

File: ci-jobs/templates/ios-e2e-template.yml [hunk 4]
```diff
@@ -70,6 +73,7 @@ jobs:
           --port=$APPIUM_TEST_SERVER_PORT \
           --address=$APPIUM_TEST_SERVER_HOST \
           --relaxed-security \
+          --keep-alive-timeout 1200 \
```
Comment (rerorero) [inline ci-jobs/templates/ios-e2e-template.yml:76]: The default is 600. It turned out that `POST /session` takes longer[ with Xcode 14.2 + iOS 16. ](https://dev.azure.com/AppiumCI/7a2a2919-d2b4-4720-a6c3-3a639f9bb2e0/_apis/build/builds/24486/logs/133). E.g. boot time gets 1.2x - 1.5x slower than Xcode 13 + iOS 15
> UDID FF980442-1833-4484-8258-3B44A8458870 booted in 333.876s

Now it takes more than 10 minutes to create a new session and test failed due to disconnection from the server (socket hang up)
```
2023-01-05T05:55:16.5477380Z [38;5;0m[HTTP][0m [37m-->[39m [37mPOST[39m [37m/session[39m
2023-01-05T06:05:16.1550910Z [38;5;0m[HTTP][0m <-- POST /session - - ms - -
```
```
2023-01-05T06:05:16.1856360Z [90m2023-01-05T06:05:16.178Z[39m [31mERROR[39m [97mwebdriver[39m: RequestError: socket hang up
2023-01-05T06:05:16.1958370Z     at ClientRequest.<anonymous> (file:///Users/runner/work/1/s/node_modules/webdriver/node_modules/got/dist/source/core/index.js:789:107)
```
So I extended the timeout.

**Comment by rerorero** [comment]
[Now only one test failed](https://dev.azure.com/AppiumCI/7a2a2919-d2b4-4720-a6c3-3a639f9bb2e0/_apis/build/builds/24486/logs/125). I'm not sure if this is just a flakey though can you rerun the test and check if it's reproducible? 
```
2023-01-12T23:25:37.8729500Z [debug] [38;5;144m[RemoteDebugger][0m No applications currently connected.
2023-01-12T23:25:37.8731820Z [debug] [38;5;32m[XCUITestDriver@e257 (4ac2e881)][0m No web frames found.
2023-01-12T23:25:37.8735130Z [38;5;32m[XCUITestDriver@e257 (4ac2e881)][0m Could not find any webviews yet, refreshing/retrying
2023-01-12T23:25:38.8224450Z [debug] [38;5;32m[XCUITestDriver@e257 (4ac2e881)][0m Retrieving contexts and views
2023-01-12T23:25:38.8227740Z [debug] [38;5;32m[XCUITestDriver@e257 (4ac2e881)][0m Selecting by url: false

```

File: ci-jobs/templates/ios-e2e-template.yml [hunk 5]
```diff
@@ -48,7 +48,9 @@ jobs:
         git checkout f660049bdec9965934973052e6331119a4985096
         brew install idb-companion
         popd
-        python -m pip install --user fb-idb
+        # install idb client
+        brew install python@3.7
```
Comment (KazuCocoa) [inline ci-jobs/templates/ios-e2e-template.yml:52]: The environment already uses Py3.7 with `UsePythonVersion` in a few lines above, so can here run `pip` install directly, or `python -m pip install ....`?
Or should be via brew for the idb?

Comment (rerorero) [reply ci-jobs/templates/ios-e2e-template.yml:52]: No, I think there is no reason to install double and I just missed it. Let me fix it, thank you.

Comment (rerorero) [reply ci-jobs/templates/ios-e2e-template.yml]: fixed

**Comment by rerorero** [comment]
hum, I realized `launchWithIDB` is not working for now 👁️ Let me take a look
https://dev.azure.com/AppiumCI/Appium%20CI/_build/results?buildId=24562&view=logs&j=8c3df2c1-6256-5af7-d8e2-3cbfdbc2c9ac

File: ci-jobs/templates/xcuitest-e2e-template.yml [hunk 6]
```diff
@@ -38,7 +39,6 @@ jobs:
       xcodeVersion: ${{ parameters.xcodeVersion }}
       deviceName: ${{ parameters.deviceName }}
       vmImage: ${{ parameters.vmImage }}
-      launchWithIDB: true
```
Comment (rerorero) [inline ci-jobs/templates/xcuitest-e2e-template.yml:41]: I realized I set it to the wrong one 🤦

**Review by KazuCocoa — APPROVED** [review]
thank you!

**Pull request merged**